### DISCUSSION TITLE
Can't get wasm file to bundle with Nextjs + Vercel (this is driving me nuts)

### DISCUSSION THREAD

**Discussion by fraywing** [vercel/community#1278] [Help] [answered]
### Summary

Been working on this for a few days. Trying to get the Tesseract wasm file to bundle but Vercel does not ever seem to find it no matter what I do. 

Things I've tried: 

1. `WasmChunksFixPlugin` (from Github)
2. adding the files to `vercel.json` 
3. including the files in `serverComponentsExternalPackages`
4. importing the files directly in api

### Additional information

Next 13.x 

```
[POST] /api/upload
14:59:39:00
Aborted(Error: ENOENT: no such file or directory, open '/var/task/node_modules/tesseract.js-core/tesseract-core-simd.wasm')
2023-01-16T22:59:42.340Z	754b4474-ff56-40ca-8804-4e8696bc9f41	ERROR	Uncaught Exception 	{"errorType":"RuntimeError","errorMessage":"Aborted(Error: ENOENT: no such file or directory, open '/var/task/node_modules/tesseract.js-core/tesseract-core-simd.wasm'). Build with -sASSERTIONS for more info.","stack":["RuntimeError: Aborted(Error: ENOENT: no such file or directory, open '/var/task/node_modules/tesseract.js-core/tesseract-core-simd.wasm'). Build with -sASSERTIONS for more info.","    at n (/var/task/node_modules/tesseract.js-core/tesseract-core-simd.js:18:375)","    at Ua (/var/task/node_modules/tesseract.js-core/tesseract-core-simd.js:19:317)","    at /var/task/node_modules/tesseract.js-core/tesseract-core-simd.js:20:394"]}
Unknown application error occurred
Runtime.Unknown
```

Next config: 

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  output: 'standalone',
  swcMinify: true,
  experimental: {
    appDir: true,
    esmExternals: true,
    serverComponentsExternalPackages: ['tesseract.js']
  },
  images: {
    domains: ['lh3.googleusercontent.com']
  },
  webpack(config, { isServer, dev }) {
    config.experiments = {
      asyncWebAssembly: true,
      layers: true
    };

    if (!dev && isServer) {
      config.output.webassemblyModuleFilename = 'chunks/[id].wasm';
      config.plugins.push(new WasmChunksFixPlugin());
    }

    return config;
  }
};

class WasmChunksFixPlugin {
  apply(compiler) {
    compiler.hooks.thisCompilation.tap('WasmChunksFixPlugin', (compilation) => {
      compilation.hooks.processAssets.tap(
        { name: 'WasmChunksFixPlugin' },
        (assets) =>
          Object.entries(assets).forEach(([pathname, source]) => {
            if (!pathname.match(/\.wasm$/)) return;
            compilation.deleteAsset(pathname);

            const name = pathname.split('/')[1];
            const info = compilation.assetsInfo.get(pathname);
            compilation.emitAsset(name, source, info);
          })
      );
    });
  }
}

module.exports = nextConfig;
```

vercel.json:

```
{
  "functions": {
    "pages/api/upload.ts": {
      "includeFiles": "node_modules/tesseract.js-core/tesseract-core-simd.wasm",
      "memory": 1024,
      "maxDuration": 10
    }
  }
}
```

**Comment by fraywing** [comment]
is it possible to force files to be bundled into `var/task/node_modules/`  for Vercel?

**Reply by AChangXD** [reply]
@fraywing any updates on this?

**Reply by shahzeb61518** [reply]
any update?

**Comment by spiceee** [comment, marked as answer]
@fraywing It Looks like your safest bet is copying the wasm files to the workspace somewhere (`public` is a good candidate) and then, from inside the function, reference the file with a `path.resolve` or `path.join`. That will trick the current `vercel` CLI into including the static files in the function deployment. https://github.com/vercel/next.js/discussions/32236#discussioncomment-2572290

**Reply by dvdokkum** [reply]
I'm running into trouble getting this workaround to work with the `ffmpeg-core.wasm` file that is part of the [ffmpeg.wasm](https://github.com/ffmpegwasm/ffmpeg.wasm) package. The linked thread is helpful, but I'm still not getting anywhere. @fraywing or @spiceee do y'all have a workinng code sample you could share? I'm using Next13, Node16. Here's what I'm trying...

```
const { createFFmpeg, fetchFile } = require('@ffmpeg/ffmpeg');
const fs = require('fs');
const path = require('path');

//various hacks to try to get ffmpeg to work load in vercel
const { getCreateFFmpegCore } = require('@ffmpeg/core');
fs.readFileSync(path.join(process.cwd(), 'public', 'ffmpeg-core.wasm'));
// const wasm = path.resolve(process.cwd(), 'public/ffmpeg-core.wasm');
// console.log(wasm);

const ffmpeg = createFFmpeg({ log: true });

export default async function handler(req, res) { ... }
```

I also added this after the handler just for good measure based on some other threads but still no dice...

```export const config = {
  unstable_includeFiles: [
    'node_modules/@ffmpeg/core/dist/ffmpeg-core.wasm',
    'public/ffmpeg-core.wasm',
    'ffmpeg-core.wasm',
  ],
};
```
Appreciate any help you may be able to offer... what a frustrating roadblock to run into at the deploy phase.

**Reply by spiceee** [reply]
@dvdokkum This is what is currently working for me:

`pages/api/convert.js`
```
import path from 'path';

import processFile from 'utils/convert/server';

const convert = async (req, res) => {
    if (req.method === 'POST') {
        // eslint-disable-next-line no-unused-vars
        const vercelFunctionHack = path.resolve('./public', '');
        // ^ Vercel doesn't comply with the serverless includeFiles directive
        // and their nft package only include files if it finds a path or fs
        // reference to it somewhere.

       // ...
       
       const data = await processFile({
            name: basename,
            byteArray,
            options: { metadata: JSON.parse(metadata), format: 'mp4', hash },
        });

       // ...
};

export default convert;
```

`utils/convert/server.js`
```
import { getCreateFFmpegCore } from '@ffmpeg/core';
// ^ Needed for the function to work at Vercel
import { createFFmpeg, fetchFile } from '@ffmpeg/ffmpeg';
import path from 'path';

const ffmpeg = createFFmpeg({
    log: true,
    corePath: path.join(process.cwd(), 'public', 'ffmpeg-core.js'),
    workerPath: path.join(process.cwd(), 'public', 'ffmpeg-core.worker.js'),
});

const processFile = async ({
    name,
    byteArray,
    setStatus = () => {},
    options = {},
}) => {
    !ffmpeg.isLoaded() && (await ffmpeg.load());

   // ...
```

**Reply by NateSrkn** [reply]
@spiceee Would it be possible to adapt this for the initial issue of tesseract? I've attempted to modify your example for ffmpeg but I haven't been able to achieve what you mentioned. Unfortunately there isn't a lot of resources or support for this.

**Reply by dvdokkum** [reply]
FWIW, I ended up using this [hack](https://github.com/vercel/next.js/discussions/32236#discussioncomment-4294677), which directly edits the `.nft.json` file as part of the build process.

**Reply by abdul-hamid-achik** [reply]
in case someone is having issues with this and uses sentry, it seems like sentry may be causing this too, although I am still not sure how

https://github.com/getsentry/sentry-javascript/issues/5998

I removed sentry for now for my setup and it works, since I need it I will add it back but I already confirmed that this

const moduleExports = {
  sentry: {
    excludeServerRoutes: ['/api/hello']
  }
}

works too

**Reply by cdric** [reply]
> @fraywing It Looks like your safest bet is copying the wasm files to the workspace somewhere (`public` is a good candidate) and then, from inside the function, reference the file with a `path.resolve` or `path.join`. That will trick the current `vercel` CLI into including the static files in the function deployment. [vercel/next.js#32236 (comment)](https://github.com/vercel/next.js/discussions/32236#discussioncomment-2572290)

This does not work for me.
I tried moving the file under /var/task/node_modules and even placing it on /public but Tesseract can't find it.

Can someone share code of a working solution?

**Reply by huumn** [reply]
Are you all on an older version of nextjs or does this not work? https://github.com/orgs/vercel/discussions/1278#discussioncomment-5445238

**Reply by cdric** [reply]
Oh I missed that. Yes this solve my problem. 
Thank you very much. Spent 2 full days trying to get to the bottom of this ::-)

Maybe this should be marked as the appropriate answer for the original question?

**Reply by tgdn** [reply]
@huumn this doesn't work on `13.5.3` at least. The `.wasm` is never bundled in the `.next/` directory, even in dev.

**Reply by huumn** [reply]
If the publicly documented api doesn't work, I'd file an issue.

**Reply by tgdn** [reply]
I'll do it today @huumn ! Thanks

**Comment by Neeraj-mvrk** [comment]
@fraywing You get anything on this bug? Coz I'm having the same issue for so long.

**Reply by ariburaco** [reply]
Hey, I have the same issue. If I may ask that did you find any solution?

**Comment by huumn** [comment]
I had a similar issue with `.wasm` ***and*** `.proto` files. 

At least with next `13.2.4` you can ["experimentally" configure the inclusion (and exclusion) of files vercel handles incorrectly in its attempt to optimize bundle size](https://nextjs.org/docs/advanced-features/output-file-tracing#caveats) in `next.config.js`:

```javascript
module.exports = {
  experimental: {
    outputFileTracingIncludes: {
      '/api/**/*': ['./node_modules/**/*.wasm', './node_modules/**/*.proto'],
    },
  },
}
```

**Reply by mfts** [reply]
This worked for me with a small caveat! 
I had to specify the routes where the file is required. the globs catchall throw errors 

```js
experimental: {
    outputFileTracingIncludes: {
      "/api/mupdf/*": ["./node_modules/mupdf/lib/*.wasm"],
   },
},
```

---
`next@13.5.4`  
`node@20`

**Comment by ariburaco** [comment]
Hey guys, I've tried to make the suggestions but no luck. Do you have a full guide to using tesseract.js with in the API routes?

**Comment by Re4GD** [comment]
Using pnpm turbo monorepo with @trustwallet/wallet-core WASM. Importing and using this package (sub-package in the monorepo) within a Nextjs package API routes.

Been struggling for hours to include wallet-core.wasm, kept getting this in production:

```TRPCClientError: Aborted(Error: ENOENT: no such file or directory, open '/var/task/apps/web/.next/server/chunks/wallet-core.wasm')```

next dev and next start worked perfectly locally, however it always failed in prod. Tried all sorts of stuff, adding experimental fields to next config, adding webpack configs to next config, editing nft.json files, importing init WASM to the next app etc. None of them worked. 

I realised that there is no ```apps/web/.next/server/chunks/wallet-core.wasm``` (from the error log) in my locally generated next build. So I had the idea of manually copying from ./node_modules to the ./.next directory and it worked for me. 

Had been using the turborepo defaults for Vercel Nextjs deploy:
Build: ```cd ../.. && pnpm build --filter web```
Root dir: ```/apps/web```

Adjusted them to:
Build: ```pnpm build --filter web && cp node_modules/@trustwallet/wallet-core/dist/lib/wallet-core.wasm ./apps/web/.next/server/chunks```
Output: ```/apps/web/.next```
Install: ```pnpm i```
Root dir: ```/```

Hope this helps

**Reply by ariburaco** [reply]
Thanks mate, I've switched to Google Vison API instead of tesseract.js, but I'll keep it in my mind if I face a issue like this. I hope it will be helpful for someone else. Cheers

**Comment by cdric** [comment]
I'm running into the same issue and it looks like the OP move to Goole Vision API and the problem wasn't solve. Has anyone been able to fix this issue? I would really want to keep using Terrasect with my project and don't want to migrate to another solution.
